File: src/main/java/com/github/jactor/persistence/service/GuestBookService.java
package com.github.jactor.persistence.service;

import com.github.jactor.persistence.dto.GuestBookDto;
import com.github.jactor.persistence.dto.GuestBookEntryDto;
import com.github.jactor.persistence.entity.GuestBookEntity;
import com.github.jactor.persistence.entity.GuestBookEntryEntity;
import com.github.jactor.persistence.repository.GuestBookEntryRepository;
import com.github.jactor.persistence.repository.GuestBookRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class GuestBookService {

    private final GuestBookRepository guestBookRepository;
    private final GuestBookEntryRepository guestBookEntryRepository;

    @Autowired
    public GuestBookService(GuestBookRepository guestBookRepository, GuestBookEntryRepository guestBookEntryRepository) {
        this.guestBookRepository = guestBookRepository;
        this.guestBookEntryRepository = guestBookEntryRepository;
    }

    public Optional<GuestBookDto> find(Long id) {
        return guestBookRepository.findById(id).map(GuestBookEntity::asDto);
    }

    public Optional<GuestBookEntryDto> findEntry(Long id) {
        return guestBookEntryRepository.findById(id).map(GuestBookEntryEntity::asDto);
    }

    public GuestBookDto saveOrUpdate(GuestBookDto guestBookDto) {
        GuestBookEntity guestBookEntity = new GuestBookEntity(guestBookDto);
        guestBookRepository.save(guestBookEntity);

        return guestBookEntity.asDto();
    }

    public GuestBookEntryDto saveOrUpdate(GuestBookEntryDto guestBookEntryDto) {
        GuestBookEntryEntity guestBookEntryEntity = new GuestBookEntryEntity(guestBookEntryDto);
        guestBookEntryRepository.save(guestBookEntryEntity);

        return guestBookEntryEntity.asDto();
    }
}

File: src/main/kotlin/com/github/jactor/persistence/service/GuestBookService.kt
package com.github.jactor.persistence.service

import java.util.UUID
import org.springframework.stereotype.Service
import com.github.jactor.persistence.dto.GuestBookModel
import com.github.jactor.persistence.dto.GuestBookEntryModel
import com.github.jactor.persistence.entity.GuestBookEntity
import com.github.jactor.persistence.entity.GuestBookEntryEntity
import com.github.jactor.persistence.repository.GuestBookEntryRepository
import com.github.jactor.persistence.repository.GuestBookRepository

interface GuestBookService {
    fun find(id: UUID): GuestBookModel?
    fun findEntry(id: UUID): GuestBookEntryModel?
    fun saveOrUpdate(guestBookModel: GuestBookModel): GuestBookModel
    fun saveOrUpdate(guestBookEntryModel: GuestBookEntryModel): GuestBookEntryModel
}

@Service
class DefaultGuestBookService(
    private val guestBookRepository: GuestBookRepository,
    private val guestBookEntryRepository: GuestBookEntryRepository
): GuestBookService {
    override fun find(id: UUID): GuestBookModel? {
        return guestBookRepository.findById(id)
            .map { it.asDto() }
            .orElse(null)
    }

    override fun findEntry(id: UUID):GuestBookEntryModel? {
        return guestBookEntryRepository.findById(id)
            .map { it.asDto() }
            .orElse(null)
    }

    override fun saveOrUpdate(guestBookModel: GuestBookModel): GuestBookModel {
        return guestBookRepository.save(GuestBookEntity(guestBookModel)).asDto()
    }

    override fun saveOrUpdate(guestBookEntryModel: GuestBookEntryModel): GuestBookEntryModel {
        return guestBookEntryRepository.save(GuestBookEntryEntity(guestBookEntryModel)).asDto()
    }
}
